### PR TITLE
[pickers] Replace `TValue` and `TSection` generics with `TIsRange`

### PR DESCRIPTION
**Pull request by flaviendelangle**
Replaced with #15434
Part of #14823 (step 2)

I will handle the migration guide in a follow up :+1: 

## Example

```diff
 UseFieldInternalProps<
-  PickerRangeValue,
-  RangeFieldSection,
+  true,
   TEnableAccessibleFieldDOMStructure,
   TimeRangeValidationError
 >
```

### CODE REVIEW

File: packages/x-date-pickers/src/internals/models/value.ts [hunk 1]
```diff
@@ -3,3 +3,17 @@ import { PickerValidDate } from '../../models/pickers';
...
+
+export type PickerNonNullableRangeValue = [PickerValidDate, PickerValidDate];
+
+export type InferPickerValue<TIsRange extends boolean> = TIsRange extends true
+  ? TIsRange extends false
+    ? PickerValue | PickerRangeValue
+    : PickerRangeValue
+  : PickerValue;
+
+export type InferNonNullablePickerValue<TIsRange extends boolean> = TIsRange extends true
```
Comment (LukasTy) [inline packages/x-date-pickers/src/internals/models/value.ts:15]: The fact that we need to put this logic and the `RangePosition` in the community package raises some "concern flags" in my head about the direction of the solution. 🙈

**Review by LukasTy — COMMENTED** [review]
I feel slightly unsure about these changes. 🤔 
Passing only a boolean into a generic type feels slightly "over-optimized". 🙈 

Sometimes a bit of explicitness is warranted.

Are you sure about this direction?

**Comment by flaviendelangle** [comment]
> I feel slightly unsure about these changes. 🤔

We can discuss it in the coming days.
I'll try to summarize why I think it improves the readability and the reliability of the codebase :+1:

File: scripts/x-date-pickers.exports.json [hunk 2]
```diff
@@ -106,6 +106,7 @@
   { "name": "ExportedUseClearableFieldProps", "kind": "Interface" },
   { "name": "extractValidationProps", "kind": "Variable" },
   { "name": "FieldFormatTokenMap", "kind": "TypeAlias" },
+  { "name": "FieldRangeSection", "kind": "Interface" },
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply scripts/x-date-pickers.exports.json:108]: Done :+1:

File: packages/x-date-pickers-pro/src/internals/hooks/useMultiInputFieldSelectedSections.ts [hunk 3]
```diff
@@ -55,15 +54,15 @@ export const useMultiInputFieldSelectedSections = (
 
   return {
     start: {
-      unstableFieldRef: params.unstableStartFieldRef,
+      unstableFieldRef: params.unstableStartFieldRef as React.Ref<FieldRef<false>>,
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/x-date-pickers-pro/src/internals/hooks/useMultiInputFieldSelectedSections.ts]: Nice catch thanks :pray:

**Pull request closed without merging**